File: 3344-minimize-manhattan-distances/minimize-manhattan-distances.cpp
typedef long long ll;

class Solution {
public:

    vector<ll> f1(const vector<vector<int>>& points)
    {
        vector<ll> a(points.size());
        for (ll i = 0; i < (ll)points.size(); i++) {
            ll x = points[i][0];
            ll y = points[i][1];
            a[i] = x + y;
        }
        return a;
    }

    vector<ll> f2(const vector<vector<int>>& points)
    {
        vector<ll> a(points.size());
        for (ll i = 0; i < (ll)points.size(); i++) {
            ll x = points[i][0];
            ll y = points[i][1];
            a[i] = x - y;
        }
        return a;
    }

   
    ll ans(vector<ll> s, vector<ll> d, ll idx)
    {
        // erase
        s.erase(s.begin() + idx);
        d.erase(d.begin() + idx);

        sort(s.begin(), s.end());
        sort(d.begin(), d.end());

        if (s.size() == 0) return 0;

        ll a = s.back() - s.front();
        ll b = d.back() - d.front();

        return max(a, b);
    }

    int minimumDistance(vector<vector<int>>& points)
    {
        ll n = points.size();

        ll maxsum = LLONG_MIN, minsum = LLONG_MAX;
        ll maxdiff = LLONG_MIN, mindiff = LLONG_MAX;
        ll maxsumidx = 0, minsumidx = 0, maxdiffidx = 0, mindiffidx = 0;

       
        for (ll i = 0; i < n; i++) {
            ll x = points[i][0];
            ll y = points[i][1];
            ll sum = x + y;
            ll diff = x - y;

            if (sum > maxsum) maxsum = sum, maxsumidx = i;
            if (sum < minsum) minsum = sum, minsumidx = i;

            if (diff > maxdiff) maxdiff = diff, maxdiffidx = i;
            if (diff < mindiff) mindiff = diff, mindiffidx = i;
        }

        vector<ll> s = f1(points);
        vector<ll> d = f2(points);

     
        ll a1 = ans(s, d, maxsumidx);
        ll a2 = ans(s, d, minsumidx);
        ll a3 = ans(s, d, maxdiffidx);
        ll a4 = ans(s, d, mindiffidx);

       
        sort(s.begin(), s.end());
        sort(d.begin(), d.end());
        ll a5 = max(s.back() - s.front(), d.back() - d.front());

        ll answer = min({a1, a2, a3, a4, a5});
        return (int)answer;
    }
};

File: 3344-minimize-manhattan-distances/minimize-manhattan-distances.py
class Solution(object):
    def f1(self, points):
        a = [0] * len(points)
        for i in range(len(points)):
            x = points[i][0]
            y = points[i][1]
            a[i] = x + y
        return a

    def f2(self, points):
        a = [0] * len(points)
        for i in range(len(points)):
            x = points[i][0]
            y = points[i][1]
            a[i] = x - y
        return a

    def ans(self, s, d, idx):
        s = s[:idx] + s[idx+1:]
        d = d[:idx] + d[idx+1:]
        s.sort()
        d.sort()
        if not s:
            return 0
        a = s[-1] - s[0]
        b = d[-1] - d[0]
        return max(a, b)

    def minimumDistance(self, points):
        n = len(points)
        maxsum = -10**30
        minsum = 10**30
        maxdiff = -10**30
        mindiff = 10**30
        maxsumidx = minsumidx = maxdiffidx = mindiffidx = 0

        for i in range(n):
            x = points[i][0]
            y = points[i][1]
            s = x + y
            d = x - y
            if s > maxsum:
                maxsum = s
                maxsumidx = i
            if s < minsum:
                minsum = s
                minsumidx = i
            if d > maxdiff:
                maxdiff = d
                maxdiffidx = i
            if d < mindiff:
                mindiff = d
                mindiffidx = i

        s = self.f1(points)
        d = self.f2(points)

        a1 = self.ans(s, d, maxsumidx)
        a2 = self.ans(s, d, minsumidx)
        a3 = self.ans(s, d, maxdiffidx)
        a4 = self.ans(s, d, mindiffidx)

        s_sorted = sorted(s)
        d_sorted = sorted(d)
        a5 = max(s_sorted[-1] - s_sorted[0], d_sorted[-1] - d_sorted[0])

        return min(a1, a2, a3, a4, a5)
